File: codes/php/polymorphism/type_manip.php
<?php

class Person {
    public function talk(){ echo "I am talking\n"; }
}

class Student extends Person {
    public function learn(){ echo "I am learning\n"; }
}

interface Worker {}

class Professor extends Person implements Worker {
    public function teach(){ echo "I am teaching\n"; }
}

$pe = new Person();
$st = new Student();
$pr = new Professor();

echo "==========================\n";
echo "Object type\n";
echo "==========================\n";
echo "gettype(pe): " . gettype($pe) . ", get_class(pe): " . get_class($pe) . "\n";
echo "gettype(st): " . gettype($st) . ", get_class(st): " . get_class($st) . "\n";
echo "gettype(pr): " . gettype($pr) . ", get_class(pr): " . get_class($pr) . "\n";
echo "gettype(25): " . gettype(25) . "\n";

echo "==========================\n";
echo "Instance of\n";
echo "==========================\n";
echo "pe instanceof 'Person': " . ($pe instanceof Person);
echo ", is_a(pe, 'Person'): " . is_a($pe, 'Person') . "\n";
echo "st instanceof Person: " . ($st instanceof Person);
echo ", is_a(st, 'Person'): " . is_a($st, 'Person') . "\n";
echo "st instanceof Professor: " . ($st instanceof Professor);
echo ", is_a(st, 'Professor'): " . is_a($st, 'Professor') . "\n";
echo "pr instanceof Professor: " . ($pr instanceof Professor);
echo ", is_a(pr, 'Professor'): " . is_a($pr, 'Professor') . "\n";
echo "pr instanceof Worker: " . ($pr instanceof Worker);
echo ", is_a(pr, 'Worker'): " . is_a($pr, 'Worker') . "\n";

echo "==========================\n";
echo "Member existance\n";
echo "==========================\n";

function testMembers($msg, $obj) {
    echo "$msg: ";
    if (is_object($obj) && method_exists($obj, "learn")) {$obj->learn();}
    else { echo "Method learn does not exist!\n";}

    echo "$msg: ";
    if (is_object($obj) && property_exists($obj, "nbr")) {
        echo "Field nbr = $obj->{nbr}\n";
    }
    else { echo "Field nbr does not exist!\n";}
}

testMembers("Person", $pe);
testMembers("Student", $st);
testMembers("Professor", $pr);
testMembers("a number", 25);

?>

File: codes/ruby/polymorphism/type_manip.rb
#!/usr/bin/ruby -w

class Person
    def talk; puts "I am talking" end
end

class Student < Person
    def learn; puts "I am learning" end
end

class Professor < Person
    attr_reader :nbr
    def initialize; @nbr = 5 end
    def teach; puts "I am teaching" end
end

module LearnerMixin
    def learn; puts "I am a learner" end
end

class Robot
    include LearnerMixin
end

pe = Person.new
st = Student.new
pr = Professor.new
ro = Robot.new

puts "=========================="
puts "Object type"
puts "=========================="
puts "pe.class: " + pe.class.to_s
puts "st.class: " + st.class.to_s
puts "pr.class: " + pr.class.to_s
puts "ro.class: " + ro.class.to_s
puts "25.class: " + 25.class.to_s

puts "=========================="
puts "Instance of"
puts "=========================="
puts "pe.is_a?(Person): " + pe.is_a?(Person).to_s
puts "st.is_a?(Person): " + st.is_a?(Person).to_s
puts "st.kind_of?(Person): " + st.kind_of?(Person).to_s
puts "st.instance_of?(Person): " + st.instance_of?(Person).to_s
puts "25.is_a?(Person): " + 25.is_a?(Person).to_s
puts "st.is_a?(Student): " + st.is_a?(Student).to_s
puts "st.instance_of?(Student): " + st.instance_of?(Student).to_s
puts "st.is_a?(Professor): " + st.is_a?(Professor).to_s
puts "ro.is_a?(Robot): " + ro.is_a?(Robot).to_s
puts "ro.is_a?(Student): " + ro.is_a?(Student).to_s
puts "ro.is_a?(LearnerMixin): " + ro.is_a?(LearnerMixin).to_s
puts "ro.instance_of?(LearnerMixin): " + ro.instance_of?(LearnerMixin).to_s
puts "ro.kind_of?(LearnerMixin): " + ro.kind_of?(LearnerMixin).to_s

puts "=========================="
puts "Member existance"
puts "=========================="

def testMembers (msg, obj)
    print "#{msg}: "
    if obj.class.method_defined? :learn then obj.learn
    else puts "Method learn does not exist!" end

    print "#{msg}: "
    if obj.class.method_defined? :nbr then puts "Field nbr = #{obj.nbr}"
    else puts "Field nbr does not exist!" end
end

testMembers("Person", pe)
testMembers("Student", st)
testMembers("Professor", pr)
testMembers("Robot", ro)
testMembers("a number", 25)
